[Validator] Add `SemVer` constraint for semantic versioning validation

OskarStark opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | https://github.com/symfony/symfony-docs/pull/21162
| License       | MIT

## Description

This PR adds a new `SemVer` constraint to validate semantic versioning strings according to the [Semantic Versioning 2.0.0](https://semver.org/) specification.

## Options

### Specific Options

- `strict` (default: `true`): Whether to use strict SemVer validation (no 'v' prefix, no partial versions) or loose validation
- `min` (default: `null`): Minimum allowed version (uses `version_compare()` for comparison)
- `max` (default: `null`): Maximum allowed version (uses `version_compare()` for comparison)
- `message` (default: `'This value is not a valid semantic version.'`): Message for invalid version format
- `minMessage` (default: `'This value should be {{ min }} or more.'`): Message when version is below minimum
- `maxMessage` (default: `'This value should be {{ max }} or less.'`): Message when version exceeds maximum

### Default Options

- `groups`: The validation groups this constraint belongs to
- `payload`: Domain-specific data attached to the constraint

## Features

### Basic Validation

The constraint validates version strings with two modes:

- **Strict mode** (default): Only accepts versions that strictly follow the SemVer 2.0.0 spec (e.g., `1.2.3`, `1.0.0-alpha`, `1.2.3+build123`)
- **Loose mode**: Also accepts common variations like partial versions and 'v' prefix (e.g., `v1.2.3`, `1.2`, `1`)

### Version Range Validation

The constraint supports `min` and `max` options to validate version ranges:

```php
use Symfony\Component\Validator\Constraints as Assert;

class Package
{
    #[Assert\SemVer(
        min: '1.0.0',
        max: '2.0.0',
        minMessage: 'This version is too old, minimum required is {{ min }}',
        maxMessage: 'This version is too new, maximum allowed is {{ max }}'
    )]
    private string $version;
}
```

## Usage Examples

### Basic usage (strict mode by default):
```php
#[Assert\SemVer]
private string $version; // Valid: "1.2.3", Invalid: "v1.2.3", "1.2"
```

### Loose mode:
```php
#[Assert\SemVer(strict: false)]
private string $version; // Valid: "1.2.3", "v1.2.3", "1.2", "1"
```

### With version constraints:
```php
#[Assert\SemVer(min: '1.0.0', max: '3.0.0')]
private string $version; // Must be between 1.0.0 and 3.0.0
```

### Custom error messages:
```php
#[Assert\SemVer(
    message: 'Invalid version format',
    minMessage: 'Version must be at least {{ min }}',
    maxMessage: 'Version cannot exceed {{ max }}'
)]
```

## Implementation Details

- Uses `version_compare()` for min/max comparisons to ensure correct semantic version ordering
- Min/max values must follow the same strict/loose rules as the validated value
- Comprehensive test coverage with 99 tests covering all scenarios

OskarStark committed:
[Validator] Add SemVer constraint for semantic versioning validation

Add a new constraint to validate semantic versioning strings according to
the SemVer 2.0.0 specification. Supports partial versions (e.g., "3", "3.1"),
full versions (e.g., "3.1.2"), optional "v" prefix, pre-release versions
(e.g., "3.1.2-beta"), and build metadata (e.g., "3.1.2+20130313144700").

Configurable options:
- requirePrefix: Enforce the "v" prefix
- allowPreRelease: Allow pre-release versions (default: true)
- allowBuildMetadata: Allow build metadata (default: true)

OskarStark committed:
Add newlines at EOF and improve code formatting

- Add missing newlines at end of files
- Break up long regex patterns with inline comments for better readability
- Convert PHPDoc @dataProvider annotations to PHP 8 attributes

OskarStark committed:
Import DataProvider attribute and use short syntax

OskarStark committed:
fix

OskarStark committed:
-

OskarStark committed:
Fix SemVer validation pattern and tests

- Fix LOOSE_SEMVER_PATTERN to only allow pre-release and build metadata with full version (major.minor.patch)
- Remove empty string from invalid test cases as it's handled separately
- Revert to PHPDoc @dataProvider annotations for PHPUnit 9.6 compatibility

xabbuh commented:
Is checking for a version really common enough to justify adding it to the Validator component?

OskarStark commented:
I also thought about just using regex, but as you can see, the regex(es) itself are quite complicated, so I thought it would be a nice addition.

Ofc lets wait what others say

javiereguiluz commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,111 @@
+        .'\.(?P<minor>0|[1-9]\d*)'                                       // Minor version
+        .'(?:'
+        .'\.(?P<patch>0|[1-9]\d*)'                                       // Patch version
+        .'(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)'  // Pre-release (only with full version)
```
Minor: this and the comments below are not aligned with the above comments

javiereguiluz commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,111 @@
+     * With optional "v" prefix: v1.0.0, v1.2.3-alpha
+     */
+    private const SEMVER_PATTERN = '/^'
+        .'(?P<prefix>v)?'                                                // Optional "v" prefix
```
The spec says that `v...` is **not** a valid semver:

> Is “v1.2.3” a semantic version?
>
> No, “v1.2.3” is not a semantic version. However, prefixing a semantic version with a “v” is a common way (in English) to indicate it is a version number. Abbreviating “version” as “v” is often seen with version control. Example: git tag v1.2.3 -m "Release version 1.2.3", in which case “v1.2.3” is a tag name and the semantic version is “1.2.3”.

See https://semver.org/#is-v123-a-semantic-version

-----

Should we add a config option to enable strict validation following the spec?

OskarStark replied:
I like the `strict` option idea

OskarStark replied:
implemented

OskarStark committed:
Simplify SemVer constraint to use single 'strict' option

- Replace requirePrefix, allowPreRelease, and allowBuildMetadata with a single 'strict' boolean option
- When strict=true: follows official SemVer spec (no 'v' prefix, requires full version)
- When strict=false: allows common variations (partial versions, 'v' prefix)
- Update tests to reflect the new behavior

OskarStark commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,82 @@
+        .'\.(?P<minor>0|[1-9]\d*)'                                       // Minor version
+        .'(?:'
+        .'\.(?P<patch>0|[1-9]\d*)'                                       // Patch version
+        .'(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)'  // Pre-release (only with full version)
```
```suggestion
        .'(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)'   // Pre-release (only with full version)
```

OskarStark committed:
Update src/Symfony/Component/Validator/Constraints/SemVerValidator.php

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/SemVer.php:
```diff
@@ -0,0 +1,49 @@
+     * @param string[]|null $groups
+     */
+    public function __construct(
+        ?array $options = null,
```
Do we want to allow configuring this new constraint with array? It's deprecated and removed in 8.0, let's not support it at all here I'd say

OskarStark replied:
you are right

OskarStark replied:
done

OskarStark committed:
Change default value of strict option to true

- Default behavior now follows the official SemVer specification
- Users must explicitly set strict=false to allow loose validation

OskarStark committed:
Remove array-based configuration support from SemVer constraint

- Remove off on off off off off off off off off on off on off on off off off on off off off on on off off off on on off off off off off off off on off off off off on off on off off off off off on off on on off off off on off on on off on on off off on on on off on on off on off off off off on off off off on off off on off on off off off on on off on off on off off on off off off on off off off off off off off off on off on off on on on off on on off off on off on on on off on on off on off on off off off off off on on off off on off off off off off on off off on on off on off off on off off off on off off off off on on off on off off off off off on off on off off off off off off off off off off on on off on off off off parameter from constructor as array-based configuration is deprecated in Symfony 8.0
- Only support named arguments for configuration
- This follows the modern Symfony constraint pattern

OskarStark committed:
Remove unnecessary comment

stof commented on src/Symfony/Component/Validator/Constraints/SemVer.php:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * @param string[]|null $groups
+     */
+    public function __construct(
```
this needs the `HasNamedArgument` attribute to be compatible with Xml and Yaml mapping files.

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,81 @@
+        .'(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?'  // Build metadata (only with full version)
+        .')?'
+        .')?'
+        .'$/';
```
You can use `x` modifier to write regex on multiple lines and add comments:

```php
private const STRICT_SEMVER_PATTERN = '/^
    (?P<major>0|[1-9]\d*)                          # Major version
    \.
    (?P<minor>0|[1-9]\d*)                          # Minor version
    \.
    (?P<patch>0|[1-9]\d*)                          # Patch version
    (?:                                            # Optional pre-release
        -
        (?P<prerelease>
            (?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*) # First segment
            (?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))* # Additional segments
        )
    )?
    (?:                                            # Optional build metadata
        \+
        (?P<buildmetadata>
            [0-9a-zA-Z-]+
            (?:\.[0-9a-zA-Z-]+)*
        )
    )?
$/x';

private const LOOSE_SEMVER_PATTERN = '/^
    (?P<prefix>v)?                                  # Optional "v" prefix
    (?P<major>0|[1-9]\d*)                           # Major version
    (?:                                             # Optional minor + patch + optional suffixes
        \.
        (?P<minor>0|[1-9]\d*)                       # Minor version
        (?:
            \.
            (?P<patch>0|[1-9]\d*)                   # Patch version
            (?:                                     # Optional pre-release
                -
                (?P<prerelease>
                    (?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)
                    (?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*
                )
            )?
            (?:                                     # Optional build metadata
                \+
                (?P<buildmetadata>
                    [0-9a-zA-Z-]+
                    (?:\.[0-9a-zA-Z-]+)*
                )
            )?
        )?
    )?
$/x';
```

OskarStark replied:
nice idea

stof commented on src/Symfony/Component/Validator/Constraints/SemVer.php:
```diff
@@ -0,0 +1,47 @@
+    ];
+
+    public string $message = 'This value is not a valid semantic version.';
+    public bool $strict = true;
```
those should be moved to default values of arguments, making them non-nullable.

fabpot replied:
Why was this comment ignored?

OskarStark replied:
By accident, on it, sorry

OskarStark replied:
Done

OskarStark committed:
Use x modifier for regex patterns to improve readability

- Convert regex patterns to multi-line format with x modifier
- Add inline comments to explain each part of the pattern
- Makes complex regex patterns more maintainable

OskarStark committed:
Apply stof's suggestions for better compatibility

- Add HasNamedArguments attribute for XML/YAML mapping compatibility
- Move property defaults to constructor arguments as non-nullable
- Follow modern Symfony constraint pattern

OskarStark commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,101 @@
+            -
+            (?P<prerelease>
+                (?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)                  # Pre-release identifier
+                (?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*          # Additional dot-separated identifiers
```
```suggestion
                (?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*           # Additional dot-separated identifiers
```

OskarStark commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,101 @@
+                    -
+                    (?P<prerelease>
+                        (?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)          # Pre-release identifier
+                        (?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*  # Additional identifiers
```
```suggestion
                        (?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*   # Additional identifiers
```

OskarStark committed:
Apply suggestions from code review

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/SemVer.php:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Validates that a value is a valid semantic version.
+ *
+ * @author Oskar Stark <oskarstark@googlemail.com>
```
```suggestion
 * Validates that a value is a valid semantic version.
 *
 * @see https://semver.org
 *
 * @author Oskar Stark <oskarstark@googlemail.com>
```

Maybe?

GromNaN reviewed (commented):
Would it be useful to add the `$min` and `$max` parameters using `version_compare` to check the value?

OskarStark commented:
Will do it tomorrow 🙌

OskarStark committed:
Update src/Symfony/Component/Validator/Constraints/SemVer.php

Co-authored-by: Alexandre Daubois <2144837+alexandre-daubois@users.noreply.github.com>

OskarStark committed:
Add min/max version constraints to SemVer validator

- Add min and max parameters to SemVer constraint
- Implement version comparison logic using version_compare()
- Add normalizeVersion() method to handle version normalization
- Add comprehensive test coverage for min/max functionality
- Validate min/max parameters follow strict mode rules

OskarStark commented on src/Symfony/Component/Validator/Tests/Constraints/SemVerValidatorTest.php:
```diff
@@ -0,0 +1,320 @@
+        $constraint = new SemVer(strict: true, min: 'v1.0.0');
+        $this->validator->validate('2.0.0', $constraint);
+    }
+
```
```suggestion
```

OskarStark committed:
Rename tooHighMessage/tooLowMessage to minMessage/maxMessage and make strict parameter required in data providers

OskarStark commented on src/Symfony/Component/Validator/Tests/Constraints/SemVerValidatorTest.php:
```diff
@@ -0,0 +1,320 @@
+        $constraint = new SemVer(strict: true, min: 'v1.0.0');
+        $this->validator->validate('2.0.0', $constraint);
+    }
+
```
```suggestion
```

OskarStark committed:
Update src/Symfony/Component/Validator/Tests/Constraints/SemVerValidatorTest.php

OskarStark committed:
Move expectException calls directly before the code that triggers the exception

OskarStark committed:
Add tests for custom message, minMessage and maxMessage

OskarStark committed:
Follow Symfony constraint pattern: initialize message properties with default values

GromNaN reviewed (approved):
Excellent!

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,166 @@
+            return;
+        }
+
+        // Normalize the version for comparison (remove 'v' prefix if present)
```
I'd remove this comment

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,166 @@
+        // Normalize the version for comparison (remove 'v' prefix if present)
+        $normalizedValue = $this->normalizeVersion($value);
+
+        // Check min constraint
```
Ditto

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,166 @@
+            }
+        }
+
+        // Check max constraint
```
Ditto

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,166 @@
+     */
+    private function normalizeVersion(string $version): string
+    {
+        // Remove 'v' prefix if present
```
Ditto

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,166 @@
+        // Remove 'v' prefix if present
+        $version = ltrim($version, 'v');
+        
+        // Split into parts
```
Ditto

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,166 @@
+        // Split into parts
+        $parts = explode('.', explode('-', explode('+', $version)[0])[0]);
+        
+        // Ensure we have at least 3 parts for version_compare
```
Ditto

alexandre-daubois commented on src/Symfony/Component/Validator/Tests/Constraints/SemVerValidatorTest.php:
```diff
@@ -0,0 +1,363 @@
+
+    public static function getInvalidStrictSemVersions(): iterable
+    {
+        // Versions with v prefix (not allowed in strict mode)
```
```suggestion
        // Versions with v prefix
```

(Self explanatory in the data provider name)

alexandre-daubois commented on src/Symfony/Component/Validator/Tests/Constraints/SemVerValidatorTest.php:
```diff
@@ -0,0 +1,363 @@
+        yield ['v1.0.0-alpha'];
+        yield ['v1.0.0+20130313144700'];
+        
+        // Partial versions (not allowed in strict mode)
```
```suggestion
        // Partial versions
```

(Self explanatory in the data provider name)

OskarStark committed:
Update src/Symfony/Component/Validator/Tests/Constraints/SemVerValidatorTest.php

Co-authored-by: Alexandre Daubois <2144837+alexandre-daubois@users.noreply.github.com>

OskarStark committed:
Update src/Symfony/Component/Validator/Tests/Constraints/SemVerValidatorTest.php

Co-authored-by: Alexandre Daubois <2144837+alexandre-daubois@users.noreply.github.com>

OskarStark committed:
Remove unnecessary comments as requested in code review

As per Alex Daubois's review comments, removed explanatory comments
that were deemed unnecessary for clean, self-documenting code.

fabpot commented on src/Symfony/Component/Validator/Constraints/SemVer.php:
```diff
@@ -0,0 +1,67 @@
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * Validates that a value is a valid semantic version.
```
Do we need to add something about the min/max contraints?

OskarStark replied:
Added the available options in L22:25.

Good to go @fabpot ?

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,160 @@
+        
+        $parts = explode('.', explode('-', explode('+', $version)[0])[0]);
+        
+        while (count($parts) < 3) {
```
```suggestion
        while (\count($parts) < 3) {
```

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/SemVerValidator.php:
```diff
@@ -0,0 +1,160 @@
+            $suffix = $matches[1];
+        }
+        
+        return implode('.', array_slice($parts, 0, 3)) . $suffix;
```
```suggestion
        $suffix = preg_match('/^[^-+]+(.+)$/', $version, $matches) ? $matches[1] : '';
        
        return implode('.', array_slice($parts, 0, 3)).$suffix;
```

nicolas-grekas commented on src/Symfony/Component/Validator/Tests/Constraints/SemVerValidatorTest.php:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    /**
+     * @dataProvider getValidLooseSemVersions
```
data providers should be put just after their consuming test case

OskarStark replied:
Can we enable https://cs.symfony.com/doc/rules/php_unit/php_unit_data_provider_method_order.html if not done yet?

OskarStark committed:
Apply code review feedback from fabpot and nicolas-grekas

- Add backslash to count() function for consistency
- Refactor regex matching to use ternary operator for more concise code
- Note: Data providers were already correctly placed after their test methods
- Note: Min/max constraints implementation is complete and working as expected

OskarStark committed:
Add documentation for SemVer constraint options

Document the available options (strict, min, max) in the class header
to make it clearer what configuration options are available.

OskarStark committed:
Refactor SemVer constraint to use non-nullable constructor arguments

- Move default values from property declarations to constructor arguments
- Make message, minMessage, maxMessage, and strict parameters non-nullable
- Rearrange parameters in logical order: strict, message, min, minMessage, max, maxMessage

OskarStark committed:
Fix parameter order: keep nullable parameters at the end

Reorder constructor parameters to follow Symfony conventions where
nullable parameters should come after non-nullable ones.

OskarStark commented:
> Note that various package managers have different regexps for versions they accept. Composer and npm don't agree on the regexp for instance.
And the regex used here does not match the Composer one (I haven't checked whether it matches the npm one)

I would say its fine to have a constraint based on the official RFC, I guess there are more projects with different regexes out there and we can't fulfill all of them

fabpot commented:
> > Note that various package managers have different regexps for versions they accept. Composer and npm don't agree on the regexp for instance.
> > And the regex used here does not match the Composer one (I haven't checked whether it matches the npm one)
> 
> I would say its fine to have a constraint based on the official RFC, I guess there are more projects with different regexes out there and we can't fulfill all of them

But then, the usefulness for this in core is limited? I suppose that supporting at least Composer would make sense?

OskarStark commented:
WDYT @alexandre-daubois @javiereguiluz @GromNaN as you approved this PR

GromNaN commented:
There may be more validations based on composer version and constraint parser. In that case, you would need to use `composer/semver`.

But I must admit that I can't think of a use for it.

OskarStark commented:
> What are the actual use cases for using that constraint ?

We release software for pur clients based on SemVer and following the official specs, and in our database we want to track when, what was delivered. So I need this constraint in this way. I don't need the min/max thing and we only use strict, so these options are only added by me for the community. OFC I can use this constraint on my own in userland.
